## Importing dependencies

In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [ ]:
import warnings
!

## Loading the dataset

In [11]:
# https://www.kaggle.com/datasets/johnsmith88/heart-disease-dataset
heart_df = pd.read_csv('heart_disease_data.csv')
heart_df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1


- age
- sex
- chest pain type (4 values)
- resting blood pressure
- serum cholestoral in mg/dl
- fasting blood sugar > 120 mg/dl
- resting electrocardiographic results (values 0,1,2)
- maximum heart rate achieved
- exercise induced angina
- oldpeak = ST depression induced by exercise relative to rest
- the slope of the peak exercise ST segment
- number of major vessels (0-3) colored by flourosopy
- thal: 0 = normal; 1 = fixed defect; 2 = reversable defect

In [12]:
heart_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [13]:
heart_df.shape

(303, 14)

In [14]:
heart_df['target'].value_counts()

target
1    165
0    138
Name: count, dtype: int64

## Splitting data into features and labels

In [16]:
X = heart_df.drop(['target'], axis=1)
Y = heart_df['target']

In [27]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=2, stratify=Y)

## Loading and Training the model

In [42]:
model = LogisticRegression(max_iter=100)
model.fit(X_train, Y_train)

D:\Program Files\anaconda\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [43]:
# Accuracy on training data
training_pred = model.predict(X_train)
score = metrics.accuracy_score(training_pred, Y_train)
print("Accuracy of the model on the training data: ", score)

Accuracy of the model on the training data:  0.8512396694214877


In [44]:
# Accuracy on test data
test_pred = model.predict(X_test)
score = metrics.accuracy_score(test_pred, Y_test)
print("Accuracy of the model on the test data: ", score)

Accuracy of the model on the test data:  0.819672131147541


## Making a predictive system

In [45]:
ind=90
input_data = heart_df.iloc[ind].drop(['target'])
actual_value = heart_df.iloc[ind]['target']

# Convert to numpy array and reshape
input_data_np = np.asarray(input_data)
input_data_reshaped = input_data_np.reshape(1, -1)

prediction = model.predict(input_data_reshaped)
if prediction[0] == 1:
    print("The patient doesn't have a heart disease")
else:
    print("The patient is likely to have heart disease")

print("Actual value:", actual_value)
print("Predicted value:", prediction[0])

The patient doesn't have a heart disease
Actual value: 1.0
Predicted value: 1


D:\Program Files\anaconda\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
